File: windows/Tidbits.Core/Models/LiveJoker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Tidbits.Core.Networking;

namespace Tidbits.Core.Models;

/// A2.14 — the joker. Before a round starts each table names ONE round of the night
/// as its joker, and every point it scores in that round counts double. Pure, and a
/// mirror of Swift's `LiveJoker`, because six stacks have to agree.
public static class LiveJoker
{
    /// The rounds a joker can still be played on: every round AFTER the one in
    /// progress (`current` 0-based; -1 before the night starts), minus the wager round.
    public static IReadOnlyList<LiveRoom.JokerRound> Playable(int current, IReadOnlyList<string> titles, int? wager = null)
    {
        var list = new List<LiveRoom.JokerRound>();
        for (int i = 0; i < titles.Count; i++)
        {
            if (i <= current || i == wager) continue;
            list.Add(new LiveRoom.JokerRound { Index = i, Title = string.IsNullOrEmpty(titles[i]) ? $"Round {i + 1}" : titles[i] });
        }
        return list;
    }

    /// Lock the picks for the round that is starting: the TEAMS (by name — G7) whose
    /// joker is on `round`. A pick for any other round names nothing here.
    public static HashSet<string> Played(int round, IReadOnlyDictionary<string, int> picks, Func<string, string?> teamOf)
    {
        var set = new HashSet<string>(StringComparer.Ordinal);
        foreach (var kv in picks)
            if (kv.Value == round && teamOf(kv.Key) is { Length: > 0 } name) set.Add(name);
        return set;
    }

    public static int Multiplier(string team, IReadOnlySet<string> played) => played.Contains(team) ? 2 : 1;

    /// The big-screen line on the first question of a round — null when nobody played one.
    public static string? Line(IEnumerable<string> played)
    {
        var names = played.OrderBy(n => n, StringComparer.Ordinal).ToList();
        if (names.Count == 0) return null;
        return (names.Count == 1 ? "Joker played: " : "Jokers played: ") + string.Join(", ", names);
    }
}

File: TidbitsTrivia/Core/Models/LiveJoker.swift
import Foundation

/// A2.14 — the joker. The oldest trick in the pub quiz: before a round starts each
/// table names ONE round of the night as its joker, and every point it scores in
/// that round counts double. It is the one decision of the night the TABLE makes
/// rather than the host — a small act of strategy that makes a table talk.
///
/// Pure, because six stacks have to agree: which rounds a joker can still be played
/// on, whose joker locks when a round starts, and what a locked joker is worth.
nonisolated enum LiveJoker {
    /// The rounds a joker can still be played on: every round AFTER the one in
    /// progress (`current` is its 0-based index; -1 before the night starts), minus
    /// the wager round — a table already stakes its own points there.
    static func playable(current: Int, titles: [String], wager: Int? = nil) -> [LiveRoom.JokerRound] {
        titles.enumerated().compactMap { i, t in
            guard i > current, i != wager else { return nil }
            return LiveRoom.JokerRound(index: i, title: t.isEmpty ? "Round \(i + 1)" : t)
        }
    }

    /// Lock the picks for the round that is starting: the TEAMS (by name, because a
    /// table may be several phones — G7) whose joker is on `round`. Picks for any
    /// other round are simply not in the set, which is the whole rule: a joker names
    /// a round before it starts, and a pick made after that names nothing.
    static func played(on round: Int, picks: [String: Int], teamOf: (String) -> String?) -> Set<String> {
        Set(picks.compactMap { uid, r in r == round ? teamOf(uid) : nil })
    }

    /// What a team's points are worth this round.
    static func multiplier(team: String, played: Set<String>) -> Int {
        played.contains(team) ? 2 : 1
    }

    /// The big-screen line on the first question of a round — nil when nobody played
    /// one, so a quiet round says nothing (A8.7).
    static func line(played: Set<String>) -> String? {
        guard !played.isEmpty else { return nil }
        let names = played.sorted()
        return (names.count == 1 ? "Joker played: " : "Jokers played: ") + names.joined(separator: ", ")
    }
}
